app/controllers/users_controller.rb

```diff
@@ -50,6 +56,55 @@ def require_login
   end
   private :require_login
 
+  def check_passive_login
+    if !user_signed_in?
+      redirect_to nyu_shibboleth_omniauth_authorize_path and return if shib_session_exists?
+      if !cookies[:_check_passive_shibboleth]
+        cookies[:_check_passive_shibboleth] = true
+        redirect_to passive_shibboleth_url
+      end
+    end
+  end
+
+  def shib_session_exists?
+    !cookies.detect {|k,v| k.include? "_shibsession_" }.nil?
+  end
+
+  def doorkeeper_client
+    @doorkeeper_client ||= Doorkeeper::Application.all.select{ |app| app.uid == params[:client_id] }.first
+  end
+
+  def doorkeeper_client_uri
+    @doorkeeper_client_uri ||= URI.parse(doorkeeper_client.redirect_uri)
+  end
+
+  def doorkeeper_client_login
+    return URI.join(doorkeeper_client_uri, params[:login_path]) if params[:login_path]
+    URI.join(doorkeeper_client_uri, "/login")
+  end
+
+  def return_uri
+    @return_uri ||= URI.parse(params[:return_uri])
+  end
+
+  def return_uri_base
+    URI.join(return_uri, "/")
+  end
+
+  def doorkeeper_client_uri_base
+    URI.join(doorkeeper_client_uri, "/")
+  end
+
+  def return_uri_validated?
+    doorkeeper_client && doorkeeper_client_uri_base.eql?(return_uri_base)
+  end
+
+  def check_passive
```

Comment:
@hab278 Can you make this and the above shibboleth passive check line-by-line `if...elsif...else` for readability?